nemo/export/trt_llm/nemo/nemo.py

```diff
@@ -0,0 +1,283 @@
+        "rotary_base": "rotary_base",
+        "rotary_scaling": "seq_len_interpolation_factor",
+        "position_embedding_type": "position_embedding_type",
+        "bias": "bias",
+        "intermediate_size": "ffn_hidden_size",
+        "num_kv_heads": "num_query_groups",
+        "moe_num_experts": "num_moe_experts",
+        "moe_top_k": "moe_router_topk",
+        "moe_renorm_mode": "moe_renorm_mode",
+        "kv_channels": "kv_channels",
+        "norm_epsilon": "layernorm_epsilon",
+    }
+
+    kwargs = {key: nemo_model_config[value] for key, value in convertion_dict.items() if value in nemo_model_config}
+    kwargs["vocab_size"] = vocab_size
+    kwargs["eos_token_id"] = eos_id
+    kwargs["bos_token_id"] = eos_id if decoder_type == 'falcon' else bos_id  # in HF falcon eos==bos
+    if "moe_num_experts" not in kwargs:
+        kwargs["moe_num_experts"] = 0
+    config_dict = {"llama": LlamaConfig, "falcon": FalconConfig, "gemma": LlamaConfig}
+    llm_config = config_dict[decoder_type] if decoder_type in config_dict else GPT2Config
+
+    return llm_config(**kwargs)
+
+
+def add_special_tokens_to_tokenizer(tokenizer):
+    # Need to add cls, sep, mask tokens to the tokenizer if they don't exist.
+    # If cls, sep and mask are not attributes of the tokenizer, add it.
+    if not hasattr(tokenizer, "cls_token"):
+        tokenizer.add_special_tokens({"cls_token": "<cls>"})
+    if not hasattr(tokenizer.tokenizer, "sep_id"):
+        tokenizer.add_special_tokens({"sep_token": "<sep>"})
+    if not hasattr(tokenizer.tokenizer, "mask_id"):
+        tokenizer.add_special_tokens({"mask_token": "<mask>"})
+
+    # bos, eos, pad and unk may be present in the provided spm .model file, if they are, use it.
+    if not hasattr(tokenizer, "pad_token"):
+        if hasattr(tokenizer.tokenizer, "pad_id") and tokenizer.tokenizer.pad_id() > 0:
+            tokenizer.pad_token = tokenizer.tokenizer.id_to_piece(tokenizer.tokenizer.pad_id())
+        else:
+            tokenizer.add_special_tokens({"pad_token": "<pad>"})
+    else:
+        tokenizer.add_special_tokens({"pad_token": "<pad>"})
+
+    if not hasattr(tokenizer, "bos_token"):
+        if hasattr(tokenizer.tokenizer, "bos_id") and tokenizer.tokenizer.bos_id() > 0:
+            tokenizer.bos_token = tokenizer.tokenizer.id_to_piece(tokenizer.tokenizer.bos_id())
+        else:
+            tokenizer.add_special_tokens({"bos_token": "<bos>"})
+    else:
+        tokenizer.add_special_tokens({"bos_token": "<s>"})
+
+    if not hasattr(tokenizer, "eos_token"):
+        if hasattr(tokenizer.tokenizer, "eos_id") and tokenizer.tokenizer.eos_id() > 0:
+            tokenizer.eos_token = tokenizer.tokenizer.id_to_piece(tokenizer.tokenizer.eos_id())
+        else:
+            tokenizer.add_special_tokens({"eos_token": "<eos>"})
+    else:
+        tokenizer.add_special_tokens({"eos_token": "</s>"})
+
+
+# TODO: remove tar.extractall usage before releasing with KitMaker
+def unpack_nemo_ckpt(
```

Comment:
If youre going o import nemo, why not import SaveRestoreConnector here and use its methods rather than copy paste them here?

Reply:
This is because of a NIM requirement. The whole export to trt-llm part of the code needs to be self contained because the NIM container doesn't want to install the whole nemo and its dependencies. That's why we had to copy and paste it. I also don't like it btw.